### PR TITLE
Replaced hard coded "new" object creation with Magento friendly way

### PR DESCRIPTION
**Pull request by woutersamaey**
Replaced hard coded "new" object creation with Magento friendly way, allowing other modules to rewrite these objects.

### CODE REVIEW

**Comment by Flyingmana** [comment]
Please use @deprecated only for classes, which are intended to get removed completely, not as a workaround to mark direct instanciation.

Also for reviewers: check usage is only changed, where magento is already loaded, there are probably uses where this will fail now, because magento is not loaded yet.

**Comment by woutersamaey** [comment]
So I remove the @deprecated word?

**Comment by Flyingmana** [comment]
yes, please :)

**Comment by woutersamaey** [comment]
I removed the 3 `@deprecated` words

**Comment by sreichel** [comment]
> Also for reviewers: check usage is only changed, where magento is already loaded, there are probably uses where this will fail now, because magento is not loaded yet.

I dont sse problems with changes inside `app/core/Mage`, but there schould be no calls to `Mage::` inside `lib/Varien` ...

**Comment by colinmollenhour** [comment]
Thanks for the PR!

Can you explain the use case for this in more detail?

Also, did you consider that you can override any class file by dropping classes with the same name into `app/code/*` directories (due to ordering of include paths)?

My thinking is that the `lib/Varien` files are utility libraries (akin to the Zend libs) and not really meant to be extended with new user-specific functionality like the models and blocks are, and if there are bugs or lacking features it is best to just submit fixes/improvements to the core code rather than override them. And if you must override them or can't wait for a PR to be merged there is always the copy/modify option with the include path.

**Comment by woutersamaey** [comment]
We have created a module that uses Amazon S3 as primary media storage without using the media dir on the disk. Other modules have S3 or other storage options too, but in our case the media dir is completely empty. 

To build this, we had to dig deep in Maganto and found these models to be not easily rewriteable. 

You’re right about the include path etc, but I would prefer Magento to load models in the same way, as Magento tells us to do. Why the ‘new’ keyword is used at all seams to break Magento’s own rules without valid reason. 

So I can understand your workaround, but my thinking was to have an easy and clean solution that does not need any workarounds. 

I’d even endeavor to remove all uses of ‘new’ as much as possible. Let devs rewrite any and all
models they like.

I understand Varien is a lib, but Magento always extends lib files and then uses these, while also loading them the Magento way. Some examples here are the Zend_Db classes. These (and others) are all wrapped in ‘Mage_’ classes. 

So hope you like this PR. There is always an alternative, but my goal is following best practises even when Magento forgot their own rules.

**Comment by colinmollenhour** [comment]
Thanks for the additional detail. You make a good case and I think I am more convinced than before so no objections.

As an aside, any chance your S3 module will be open-source? :)

**Comment by woutersamaey** [comment]
The S3 module actually needs more core edits, but it's very hard for me to post them here and get everyone to agree on them. So, let's first try to get this one in...

**Comment by sreichel** [comment]
Are the changes to lib/Varien required if you use your added wrapper classes?

**Comment by tmotyl** [comment]
This kind of change has a significant risk of introducing some side effects.
Maybe we are missing some api or event which would make the whole architecture of your extension clearer and you would not need to override core classes any more?

Can you describe implementation details of your change?

**Review by fballiano — DISMISSED** [review]

**Review by colinmollenhour — APPROVED** [review]

File: app/code/core/Mage/Catalog/Helper/Image.php [hunk 1]
```diff
@@ -646,7 +646,7 @@ public function validateUploadFile($filePath)
             Mage::throwException($this->__('Disalollowed file format.'));
         }
 
-        $_processor = new Varien_Image($filePath);
+        $_processor = Mage::getModel('core/varien_image',$filePath);
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Helper/Image.php:649]: You have forgotten a space after comma.

File: app/code/core/Mage/Catalog/Model/Api2/Product/Image/Rest/Admin/V1.php [hunk 2]
```diff
@@ -72,7 +72,7 @@ protected function _create(array $data)
             // try to create Image object to check if image data is valid
             try {
                 $filePath = $apiTempDir . DS . $imageFileName;
-                new Varien_Image($filePath);
+                Mage::getModel('core/varien_image',$filePath);
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Model/Api2/Product/Image/Rest/Admin/V1.php:75]: You have forgotten a space after comma.

File: app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php [hunk 3]
```diff
@@ -58,7 +58,7 @@ protected function _beforeSave()
                 $file['tmp_name'] = $tmpName[$this->getGroupId()]['fields'][$this->getField()]['value'];
                 $name = $_FILES['groups']['name'];
                 $file['name'] = $name[$this->getGroupId()]['fields'][$this->getField()]['value'];
-                $uploader = new Mage_Core_Model_File_Uploader($file);
+                $uploader = Mage::getModel('core/file_uploader',$file);
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/Model/System/Config/Backend/File.php:61]: You have forgotten a space after comma.

File: app/code/core/Mage/Adminhtml/controllers/Catalog/Product/GalleryController.php [hunk 4]
```diff
@@ -36,7 +36,7 @@ class Mage_Adminhtml_Catalog_Product_GalleryController extends Mage_Adminhtml_Co
     public function uploadAction()
     {
         try {
-            $uploader = new Mage_Core_Model_File_Uploader('image');
+            $uploader = Mage::getModel('core/file_uploader','image');
```
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/controllers/Catalog/Product/GalleryController.php:39]: You have forgotten a space after comma.

File: app/code/core/Mage/Catalog/Model/Category/Attribute/Backend/Image.php [hunk 5]
```diff
@@ -54,7 +54,7 @@ public function afterSave($object)
         $path = Mage::getBaseDir('media') . DS . 'catalog' . DS . 'category' . DS;
 
         try {
-            $uploader = new Mage_Core_Model_File_Uploader($this->getAttribute()->getName());
+            $uploader = Mage::getModel('core/file_uploader',$this->getAttribute()->getName());
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Model/Category/Attribute/Backend/Image.php:57]: You have forgotten a space after comma.

File: app/code/core/Mage/Catalog/Model/Product/Attribute/Media/Api.php [hunk 6]
```diff
@@ -161,7 +161,7 @@ public function create($productId, $data, $store = null, $identifierType = null)
             // try to create Image object - it fails with Exception if image is not supported
             try {
                 $filePath = $tmpDirectory . DS . $fileName;
-                new Varien_Image($filePath);
+                Mage::getModel('core/varien_image',$filePath);
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Model/Product/Attribute/Media/Api.php:164]: You have forgotten a space after comma.

File: app/code/core/Mage/Catalog/Model/Product/Image.php [hunk 7]
```diff
@@ -413,7 +413,7 @@ public function getImageProcessor()
 //            if (!$this->_checkMemory()) {
 //                $this->_baseFile = null;
 //            }
-            $this->_processor = new Varien_Image($this->getBaseFile());
+            $this->_processor = Mage::getModel('core/varien_image',$this->getBaseFile());
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Model/Product/Image.php:416]: You have forgotten a space after comma.

File: app/code/core/Mage/Catalog/Model/Resource/Product/Attribute/Backend/Image.php [hunk 8]
```diff
@@ -52,7 +52,7 @@ public function afterSave($object)
         }
 
         try {
-            $uploader = new Mage_Core_Model_File_Uploader($this->getAttribute()->getName());
+            $uploader = Mage::getModel('core/file_uploader',$this->getAttribute()->getName());
```
Comment (luigifab) [inline app/code/core/Mage/Catalog/Model/Resource/Product/Attribute/Backend/Image.php:55]: You have forgotten a space after comma.

File: app/code/core/Mage/Cms/Model/Wysiwyg/Images/Storage.php [hunk 9]
```diff
@@ -282,7 +282,7 @@ public function deleteFile($target)
      */
     public function uploadFile($targetPath, $type = null)
     {
-        $uploader = new Mage_Core_Model_File_Uploader('image');
+        $uploader = Mage::getModel('core/file_uploader','image');
```
Comment (luigifab) [inline app/code/core/Mage/Cms/Model/Wysiwyg/Images/Storage.php:285]: You have forgotten a space after comma.

File: app/code/core/Mage/ConfigurableSwatches/Helper/Productimg.php [hunk 10]
```diff
@@ -308,7 +308,7 @@ protected function _resizeSwatchImage($filename, $tag, $width, $height)
             }
 
             // Do resize and save
-            $processor = new Varien_Image($sourceFilePath);
+            $processor = Mage::getModel('core/varien_image',$sourceFilePath);
```
Comment (luigifab) [inline app/code/core/Mage/ConfigurableSwatches/Helper/Productimg.php:311]: You have forgotten a space after comma.

File: app/code/core/Mage/Dataflow/Model/Convert/Adapter/Http.php [hunk 11]
```diff
@@ -74,7 +74,7 @@ public function loadFile()
             exit;
         }
         if (!empty($_FILES['io_file']['tmp_name'])) {
-            $uploader = new Mage_Core_Model_File_Uploader('io_file');
+            $uploader = Mage::getModel('core/file_uploader','io_file');
```
Comment (luigifab) [inline app/code/core/Mage/Dataflow/Model/Convert/Adapter/Http.php:77]: You have forgotten a space after comma.

File: app/code/core/Mage/Dataflow/Model/Convert/Iterator/Http.php [hunk 12]
```diff
@@ -45,7 +45,7 @@ public function load()
             exit;
         }
         if (!empty($_FILES['io_file']['tmp_name'])) {
-            $uploader = new Mage_Core_Model_File_Uploader('io_file');
+            $uploader = Mage::getModel('core/file_uploader','io_file');
```
Comment (luigifab) [inline app/code/core/Mage/Dataflow/Model/Convert/Iterator/Http.php:48]: You have forgotten a space after comma.

File: app/code/core/Mage/Dataflow/Model/Profile.php [hunk 13]
```diff
@@ -179,7 +179,7 @@ protected function _afterSave()
         || isset($_FILES['file_3']['tmp_name'])) {
             for ($index = 0; $index < 3; $index++) {
                 if ($file = $_FILES['file_' . ($index+1)]['tmp_name']) {
-                    $uploader = new Mage_Core_Model_File_Uploader('file_' . ($index + 1));
+                    $uploader = Mage::getModel('core/file_uploader','file_' . ($index + 1));
```
Comment (luigifab) [inline app/code/core/Mage/Dataflow/Model/Profile.php:182]: You have forgotten a space after comma.

File: app/code/core/Mage/Dataflow/Model/Session/Adapter/Http.php [hunk 14]
```diff
@@ -46,7 +46,7 @@ public function load()
          }
 
          if (!empty($_FILES['io_file']['tmp_name'])) {
-            $uploader = new Mage_Core_Model_File_Uploader('io_file');
+            $uploader = Mage::getModel('core/file_uploader','io_file');
```
Comment (luigifab) [inline app/code/core/Mage/Dataflow/Model/Session/Adapter/Http.php:49]: You have forgotten a space after comma.

File: app/code/core/Mage/Downloadable/controllers/Adminhtml/Downloadable/FileController.php [hunk 15]
```diff
@@ -50,7 +50,7 @@ public function uploadAction()
         }
         $result = array();
         try {
-            $uploader = new Mage_Core_Model_File_Uploader($type);
+            $uploader = Mage::getModel('core/file_uploader',$type);
```
Comment (luigifab) [inline app/code/core/Mage/Downloadable/controllers/Adminhtml/Downloadable/FileController.php:53]: You have forgotten a space after comma.

File: app/code/core/Mage/Eav/Model/Attribute/Data/File.php [hunk 16]
```diff
@@ -226,15 +226,15 @@ public function compactValue($value)
...
+            $ioFile = Mage::getModel('core/varien_io_file');
             if ($ioFile->fileExists($file)) {
                 $ioFile->rm($file);
             }
         }
 
         if (!empty($value['tmp_name'])) {
             try {
-                $uploader = new Varien_File_Uploader($value);
+                $uploader = Mage::getModel('core/varien_file_uploader',$value);
```
Comment (luigifab) [inline app/code/core/Mage/Eav/Model/Attribute/Data/File.php:237]: You have forgotten a space after comma.

File: app/code/core/Mage/XmlConnect/Helper/Image.php [hunk 17]
```diff
@@ -224,7 +224,7 @@ protected function _handleResize($fieldPath, $file)
             }
         }
 
-        $image = new Varien_Image($file);
+        $image = Mage::getModel('core/varien_image',$file);
```
Comment (luigifab) [inline app/code/core/Mage/XmlConnect/Helper/Image.php:227]: You have forgotten a space after comma.

File: app/code/core/Mage/XmlConnect/Helper/Image.php [hunk 18]
```diff
@@ -484,7 +484,7 @@ public function getCustomSizeImageUrl($imageUrl, $width = 100, $height = 100)
         $file = array_pop($imageUrl);
         $filePath = Mage_XmlConnect_Model_Images::getBasePath() . DS . $file;
         if (!file_exists($customDir . DS . $file)) {
-            $image = new Varien_Image($filePath);
+            $image = Mage::getModel('core/varien_image',$filePath);
```
Comment (luigifab) [inline app/code/core/Mage/XmlConnect/Helper/Image.php:487]: You have forgotten a space after comma.

File: app/code/core/Mage/XmlConnect/Model/Images.php [hunk 19]
```diff
@@ -276,7 +276,7 @@ public function getCustomSizeImageUrl($imageFile, $width = 100, $height = 100)
                 $filePath = $this->_forcedConvertPng($filePath, $customSizeFile, $originalImageType);
             }
 
-            $image = new Varien_Image($filePath);
+            $image = Mage::getModel('core/varien_image',$filePath);
```
Comment (luigifab) [inline app/code/core/Mage/XmlConnect/Model/Images.php:279]: You have forgotten a space after comma.

File: lib/Varien/Convert/Adapter/Http.php [hunk 20]
```diff
@@ -74,7 +74,7 @@ public function loadFile()
          }
          if (!empty($_FILES['io_file']['tmp_name'])) {
             //$this->setData(file_get_contents($_FILES['io_file']['tmp_name']));
-            $uploader = new Varien_File_Uploader('io_file');
+            $uploader = Mage::getModel('core/varien_file_uploader','io_file');
```
Comment (luigifab) [inline lib/Varien/Convert/Adapter/Http.php:77]: You have forgotten a space after comma.

File: lib/Varien/Image.php [hunk 21]
```diff
@@ -26,6 +26,7 @@
 
 /**
  * Image handler library
+ * Instead of creating instances using "new", use Mage::getModel('core/image')
```
Comment (luigifab) [inline lib/Varien/Image.php:29]: That's wrong, it's: `Mage::getModel('core/varien_image')`.

File: lib/Varien/Io/File.php [hunk 22]
```diff
@@ -27,6 +27,7 @@
 
 /**
  * Filesystem client
+ * Instead of creating instances using "new", use Mage::getModel('core/io_file')
```
Comment (luigifab) [inline lib/Varien/Io/File.php:30]: That's wrong, it's: `Mage::getModel('core/varien_io_file')`.

**Comment by woutersamaey** [comment]
@luigifab I fixed these few nit-picks

**Comment by fballiano** [comment]
I love how @luigifab is our syntax police (I really do, that's not sarcastic) :-)

**Comment by woutersamaey** [comment]
We should use PHP CS Fixer for these nit-picks. Devs really have better things to do imo.

**Comment by ADDISON74** [comment]
From what I understand this PR comes as a proposal to solve a particular situation encountered in Magento. At first glance, the changes are made correctly, but what are the implications in the wrong sense?

If this PR will be merged I would change the phrase bellow is by removing the comma:

```Instead of creating instances using "new", use Mage::getModel('core/varien_file_uploader')```

**Comment by fballiano** [comment]
> Are the changes to lib/Varien required if you use your added wrapper classes?

I agree with @sreichel that it's very weird to see Mage:: in "lib/", is it absolutely necessary for some reason?

**Review by luigifab — COMMENTED** [review]
I tried this:
```php
echo get_class(new Varien_Io_File()),"\n";
echo get_class(Mage::getModel('Varien_Io_File')),"\n";
echo get_class(Mage::getModel('varien/io_file')),"\n";
```
Results are:
```
Varien_Io_File
Varien_Io_File
Varien_Io_File
```
So, `Mage_Core_Model_Varien_File_Uploader` + `Mage_Core_Model_Varien_Image` + `Mage_Core_Model_Varien_Io_File` are not necessary?

**Comment by fballiano** [comment]
I wouldn't have this in v20 only, it will create too many merging problems.

Also, there are conflicts to be fixed, I'm setting it to draft.

File: lib/Varien/File/Uploader.php [hunk 23]
```diff
@@ -30,6 +30,8 @@
  * ATTENTION! This class must be used like abstract class and must added
  * validation by protected file extension list to extended class
  *
+ * Instead of creating instances using "new", use Mage::getModel('core/varien_file_uploader')
+ *
```
[the remark this answers is not in the document]
Comment (fballiano) [reply lib/Varien/File/Uploader.php:34]: anyway files in "lib/" should never have references to "Mage::" as Sven pointed out

[the remark this answers is not in the document]
Comment (ADDISON74) [reply lib/Varien/File/Uploader.php:34]: I only wanted to correct a few sentences, This PR has some important issues that need to be discussed.

[the remark this answers is not in the document]
Comment (fballiano) [reply lib/Varien/File/Uploader.php:34]: yep for sure, if we merge luigifab PR (which is similar) I'll take this one further but I without the "lib/" part ;-)

**Pull request closed without merging**